gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StorageStubProvider.java

```diff
@@ -209,45 +198,92 @@ private synchronized ManagedChannel getManagedChannel() {
+  }
+
+  static class CloudPathGrpcDecorator implements GrpcDecorator {
+    private final Credentials credentials;
+
+    CloudPathGrpcDecorator(Credentials credentials) {
+      this.credentials = credentials;
+    }
+
+    public ManagedChannelBuilder<?> createChannelBuilder(String target) {
+      return ManagedChannelBuilder.forTarget(target);
+    }
+
+    public AbstractStub<?> applyCallOption(AbstractStub<?> stub) {
+      return stub.withCallCredentials(MoreCallCredentials.from(credentials));
+    }
+  }
+
+  static class DirectpathPathGrpcDecorator implements GrpcDecorator {
+    private final GoogleCloudStorageReadOptions readOptions;
+
+    DirectpathPathGrpcDecorator(GoogleCloudStorageReadOptions readOptions) {
+      this.readOptions = readOptions;
+    }
+
+    public ManagedChannelBuilder<?> createChannelBuilder(String target) {
+      return GoogleDefaultChannelBuilder.forTarget(target).defaultServiceConfig(getGrpcServiceConfig());
+    }
+
+    public AbstractStub<?> applyCallOption(AbstractStub<?> stub) {
+      return stub;
+    }
+
+    private Map<String, Object> getGrpcServiceConfig() {
+      Map<String, Object> name = ImmutableMap.of("service", "google.storage.v1.Storage");
+
+      Map<String, Object> retryPolicy = ImmutableMap.<String, Object>builder()
+          .put("maxAttempts", GRPC_MAX_RETRY_ATTEMPTS)
+          .put("initialBackoff",
+              Durations.toString(Durations.fromMillis(readOptions.getBackoffInitialIntervalMillis())))
+          .put("maxBackoff", Durations.toString(Durations.fromMillis(readOptions.getBackoffMaxIntervalMillis())))
+          .put("backoffMultiplier", readOptions.getBackoffMultiplier())
+          .put("retryableStatusCodes", ImmutableList.of("UNAVAILABLE", "RESOURCE_EXHAUSTED")).build();
+
+      Map<String, Object> methodConfig = ImmutableMap.of("name", ImmutableList.of(name), "retryPolicy", retryPolicy);
+
+      // When channel pooling is enabled, force the pick_first grpclb strategy.
+      // This is necessary to avoid the multiplicative effect of creating channel pool
+      // with
+      // `poolSize` number of `ManagedChannel`s, each with a `subSetting` number of
+      // number of
+      // subchannels.
+      // See the service config proto definition for more details:
+      // https://github.com/grpc/grpc-proto/blob/master/grpc/service_config/service_config.proto#L182
+      Map<String, Object> pickFirstStrategy = ImmutableMap.of("pick_first", ImmutableMap.of());
+
+      Map<String, Object> childPolicy = ImmutableMap.of("childPolicy", ImmutableList.of(pickFirstStrategy));
+
+      Map<String, Object> grpcLbPolicy = ImmutableMap.of("grpclb", childPolicy);
+
+      return ImmutableMap.of("methodConfig", ImmutableList.of(methodConfig), "loadBalancingConfig",
+          ImmutableList.of(grpcLbPolicy));
+    }
+  }
+
+  public static StorageStubProvider create(GoogleCloudStorageOptions options, ExecutorService backgroundTasksThreadPool,
```

Comment:
how about using `newInstance` instead of `create` ?

Reply:
yep, it sounds better.